akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/PoolRouterImpl.scala

```diff
@@ -42,30 +42,37 @@ private final class PoolRouterImpl[T](
   logic:    RoutingLogic[T]
 ) extends AbstractBehavior[T] {
 
-  private var routees = (1 to poolSize).map { _ ⇒
+  (1 to poolSize).map { _ ⇒
     val child = ctx.spawnAnonymous(behavior)
     ctx.watch(child)
     child
-  }.toArray
+  }
+  onRouteesChanged()
+
+  private def onRouteesChanged(): Unit = {
+    val children = ctx.children.toVector.asInstanceOf[Vector[ActorRef[T]]]
+    logic.routeesUpdated(children)
+  }
 
   def onMessage(msg: T): Behavior[T] = {
-    val recipient = logic.selectRoutee(routees)
-    recipient.tell(msg)
+    logic.selectRoutee() ! msg
     this
   }
 
   override def onSignal: PartialFunction[Signal, Behavior[T]] = {
     case Terminated(child) ⇒
-      ctx.log.warning(s"Pool child stopped [${child.path}]")
-      val childIdx = routees.indexOf(child.unsafeUpcast[T])
-      if (childIdx == -1)
-        throw new IllegalStateException(s"Got termination message for [${child.path}] which isn't a routee of this router")
-      val originalRoutees = routees
-      routees = routees.filterNot(_ == child)
-      if (routees.nonEmpty) {
-        logic.routeesUpdated(originalRoutees, routees)
+      // Note that if several children are stopping concurrently children may already be empty
+      // for the `Terminated` we receive for the first child. This means it is not certain that
+      // there will be a log entry per child in those cases (it does not make sense to keep the
+      // pool alive just to get the logging right when there are no routees available)
+      if (ctx.children.nonEmpty) {
+        ctx.log.warning("Pool child stopped [{}]", child.path)
```

Comment:
not sure that is a warning? supervision will log exceptions

Reply:
Yeah, went back and forth about this. debug for child and then a warning when/if all children stops?

Reply:
My initial motivation was that a user would miss out on supervision, good point that it will anyways log. Info for pool stopping?

Reply:
debug for child stop and info on pool stopping sounds good